### PR TITLE
Refactor previous pairs and add redo pairing functionality

### PR DESCRIPTION
**Pull request by goodyguts**
@AmaanIbnNasar there are a lot of changes in this PR so I would appreciate if you'd pull and check that the program works and is well organised.

Also if you'd code review it I'd be grateful

### CODE REVIEW

**Comment by AmaanIbnNasar** [comment]
I'll take a look when I can
Got quite a busy weekend ahead of me with house stuff so I'll do my best

**Comment by goodyguts** [comment]
we're on a tight release schedule Amaan!

File: pairModule/fileHandling.py [hunk 1]
```diff
@@ -13,35 +13,35 @@ def getPeople(team):
...
-    previousPairSets.insert(0, {
-        **{
-            p1: p2
-            for p1, p2 in pairings
-        },
-        **{
-            person: 0
-            for person in unpaired
-        }
+def savePairings(team, name, pairings, excluded, unpaired, location, *, overwrite_most_recent=False):
```
Comment (AmaanIbnNasar) [inline pairModule/fileHandling.py:30]: Not a specific code change but just wondered what the * argument means and what it's used for?

File: pairModule/pairingProcesses.py [hunk 2]
```diff
@@ -25,58 +26,114 @@ def pairAllTeams():
...
+        redo_answer = inquirer.prompt([
+            inquirer.Confirm(
+                'redo',
+                message='Would you like to ignore and redo the last pairing (displayed above)?',
+                default=False
+            )
+        ])['redo']
+
+        print(
+            "\nThe last pairing will be used to set defaults for the upcoming questions.\n")
```
Comment (AmaanIbnNasar) [inline pairModule/pairingProcesses.py:47]: Should this message change based on the redo_answer?

File: pairModule/pairingProcesses.py [hunk 3]
```diff
@@ -25,58 +26,114 @@ def pairAllTeams():
...
+    else:
+        redo_answer = False
+        print("\nNo previous pairings found.\n")
+
+    if lastPairing:
+        preselected_wfh_people = []
+        unpreselected_wfh_people = []
+        for person in people:
+            if lastPairing["location"].get(person, None) == "home":
```
Comment (AmaanIbnNasar) [inline pairModule/pairingProcesses.py:56]: Another question for my own sake
What does the None in the .get mean?

**Comment by AmaanIbnNasar** [comment]
Done some testing and I think everything works as expected
I'm sure we'll find some bugs when we do formal testing

File: pairModule/fileHandling.py (hunk 1 above)
Comment (goodyguts) [reply pairModule/fileHandling.py:30]: * means that the arguments after that * cannot be passed as positional arguments, they have to be passed as keyword arguments

Comment (goodyguts) [reply pairModule/fileHandling.py:30]: `f(5, b=7)` then the first one is a normal arg, the second one is a kwarg

File: pairModule/pairingProcesses.py (hunk 2 above)
Comment (goodyguts) [reply pairModule/pairingProcesses.py:47]: I wanted the defaults to be set whether or not people redo the thing. redo in this implementation only means that the previous pairing is ignored in the matching algorithm and it gives you the option of overwriting

File: pairModule/pairingProcesses.py (hunk 3 above)
Comment (goodyguts) [reply pairModule/pairingProcesses.py:56]: it is the default if person is not in that dictionary

**Pull request merged**